File: src/main/java/sharev/domain/team/controller/TeamController.java
package sharev.domain.team.controller;

import jakarta.validation.Valid;
import java.util.List;
import lombok.RequiredArgsConstructor;
import org.springframework.http.HttpStatus;
import org.springframework.http.ResponseEntity;
import org.springframework.security.access.prepost.PreAuthorize;
import org.springframework.security.core.annotation.AuthenticationPrincipal;
import org.springframework.web.bind.annotation.GetMapping;
import org.springframework.web.bind.annotation.PatchMapping;
import org.springframework.web.bind.annotation.PathVariable;
import org.springframework.web.bind.annotation.PostMapping;
import org.springframework.web.bind.annotation.RequestBody;
import org.springframework.web.bind.annotation.RequestMapping;
import org.springframework.web.bind.annotation.RestController;
import sharev.domain.account.entity.Account;
import sharev.domain.team.dto.request.RequestCreateTeamDto;
import sharev.domain.team.dto.request.RequestUpdateTeamDto;
import sharev.domain.team.dto.response.ResponseTeamInfoDto;
import sharev.domain.team.service.TeamService;

@RestController
@RequestMapping("/teams")
@RequiredArgsConstructor
public class TeamController {
    private final TeamService teamService;

    @PostMapping
    public ResponseEntity<Void> createTeam(@AuthenticationPrincipal Account account,
                                           @Valid @RequestBody RequestCreateTeamDto requestCreateTeamDto) {
        teamService.create(account, requestCreateTeamDto.title());
        return ResponseEntity.status(HttpStatus.CREATED)
                .build();
    }

    @GetMapping
    public ResponseEntity<List<ResponseTeamInfoDto>> getMyTeams(@AuthenticationPrincipal Account account) {
        return ResponseEntity.ok(teamService.getMyTeams(account));
    }

    // TODO: GET /{teamId} 팀 상세 정보(행사 목록, 멤버 목록, 팀명)

    @PatchMapping("/{teamId}")
    @PreAuthorize("@teamService.isMember(authentication.principal, #teamId)")
    public ResponseEntity<Void> updateTeamInfo(@PathVariable Long teamId,
                                               @RequestBody RequestUpdateTeamDto requestUpdateTeamDto) {
        teamService.updateTeamInfo(teamId, requestUpdateTeamDto.title());
        return ResponseEntity.ok()
                .build();
    }
}

File: src/main/kotlin/sharev/team/adapter/inbound/web/controller/TeamController.kt
package sharev.team.adapter.inbound.web.controller

import jakarta.validation.Valid
import org.springframework.http.HttpStatus
import org.springframework.http.ResponseEntity
import org.springframework.security.core.annotation.AuthenticationPrincipal
import org.springframework.web.bind.annotation.*
import sharev.common.adapter.inbound.security.model.AccountPrincipal
import sharev.team.adapter.inbound.web.dto.request.CreateTeamRequest
import sharev.team.adapter.inbound.web.dto.request.UpdateTeamRequest
import sharev.team.adapter.inbound.web.dto.response.CreateTeamResponse
import sharev.team.adapter.inbound.web.dto.response.TeamDetailResponse
import sharev.team.adapter.inbound.web.dto.response.TeamInfoResponse
import sharev.team.adapter.inbound.web.dto.response.TeamUpdateInfoResponse
import sharev.team.adapter.inbound.web.mapper.toCommand
import sharev.team.adapter.inbound.web.mapper.toResponse
import sharev.team.application.port.inbound.command.GetMyTeamsCommand
import sharev.team.application.port.inbound.usecase.CreateTeamUseCase
import sharev.team.application.port.inbound.usecase.GetMyTeamsUseCase
import sharev.team.application.port.inbound.usecase.GetTeamDetailUseCase
import sharev.team.application.port.inbound.usecase.UpdateTeamInfoUseCase

@RestController
@RequestMapping("/teams")
class TeamController(
    private val createTeamUseCase: CreateTeamUseCase,
    private val getMyTeamsUseCase: GetMyTeamsUseCase,
    private val getTeamDetailUseCase: GetTeamDetailUseCase,
    private val updateTeamInfoUseCase: UpdateTeamInfoUseCase,
) {

    @PostMapping
    fun createTeam(
        @AuthenticationPrincipal accountPrincipal: AccountPrincipal,
        @Valid @RequestBody request: CreateTeamRequest,
    ): ResponseEntity<CreateTeamResponse> {

        val response = createTeamUseCase.create(request.toCommand(accountPrincipal.id))
            .toResponse()

        return ResponseEntity.status(HttpStatus.CREATED)
            .body(response)
    }

    @GetMapping
    fun getMyTeams(
        @AuthenticationPrincipal accountPrincipal: AccountPrincipal,
    ): ResponseEntity<List<TeamInfoResponse>> {

        val response = getMyTeamsUseCase.getMyTeams(GetMyTeamsCommand(accountPrincipal.id))
            .map { it.toResponse() }

        return ResponseEntity.ok(response)
    }

    @GetMapping("/{teamId}")
    fun getTeamDetail(
        @AuthenticationPrincipal accountPrincipal: AccountPrincipal,
        @PathVariable teamId: Long,
    ): ResponseEntity<TeamDetailResponse> {

        val response = getTeamDetailUseCase.getTeamDetail(accountPrincipal.id, teamId)
            .toResponse()

        return ResponseEntity.ok(response)
    }

    @PatchMapping("/{teamId}")
    fun updateTeamInfo(
        @AuthenticationPrincipal accountPrincipal: AccountPrincipal,
        @PathVariable teamId: Long,
        @Valid @RequestBody request: UpdateTeamRequest,
    ): ResponseEntity<TeamUpdateInfoResponse> {

        val response = updateTeamInfoUseCase.updateTeamInfo(
            request.toCommand(
                accountPrincipal.id, teamId
            )
        ).toResponse()

        return ResponseEntity.ok(response)
    }
}
